packages/trie/test/checkpoint.spec.ts

```diff
@@ -1,180 +1,182 @@
-    const { root } = trie
-    trie.checkpoint()
-    await trie.put(Buffer.from('the feels'), Buffer.from('emotion'))
-    await trie.revert()
-    await trie.commit()
-    t.equal(trie.isCheckpoint, false)
-    t.equal(trie.root.toString('hex'), root.toString('hex'))
-    t.end()
-  })
-
-  const k1 = Buffer.from('k1')
-  const v1 = Buffer.from('v1')
-  const v12 = Buffer.from('v12')
-  const v123 = Buffer.from('v123')
-
-  it('revert -> put', async function (t) {
-    trie = new CheckpointTrie({ db: new LevelDB() })
-
-    trie.checkpoint()
-    await trie.put(k1, v1)
-    t.deepEqual(await trie.get(k1), v1, 'before revert: v1 in trie')
-    await trie.revert()
-    t.deepEqual(await trie.get(k1), null, 'after revert: v1 removed')
-
-    t.end()
-  })
-
-  it('revert -> put (update)', async (t) => {
-    trie = new CheckpointTrie({ db: new LevelDB() })
-
-    await trie.put(k1, v1)
-    t.deepEqual(await trie.get(k1), v1, 'before CP: v1')
-    trie.checkpoint()
-    await trie.put(k1, v12)
-    await trie.put(k1, v123)
-    await trie.revert()
-    t.deepEqual(await trie.get(k1), v1, 'after revert: v1')
-    t.end()
-  })
-
-  it('revert -> put (update) batched', async (t) => {
-    const trie = new CheckpointTrie({ db: new LevelDB() })
-    await trie.put(k1, v1)
-    t.deepEqual(await trie.get(k1), v1, 'before CP: v1')
-    trie.checkpoint()
-    const ops = [
-      { type: 'put', key: k1, value: v12 },
-      { type: 'put', key: k1, value: v123 },
-    ] as BatchDBOp[]
-    await trie.batch(ops)
-    await trie.revert()
-    t.deepEqual(await trie.get(k1), v1, 'after revert: v1')
-    t.end()
-  })
-
-  it('Checkpointing: revert -> del', async (t) => {
-    const trie = new CheckpointTrie({ db: new LevelDB() })
-    await trie.put(k1, v1)
-    t.deepEqual(await trie.get(k1), v1, 'before CP: v1')
-    trie.checkpoint()
-    await trie.del(k1)
-    t.deepEqual(await trie.get(k1), null, 'before revert: null')
-    await trie.revert()
-    t.deepEqual(await trie.get(k1), v1, 'after revert: v1')
-    t.end()
-  })
-
-  it('Checkpointing: nested checkpoints -> commit -> revert', async (t) => {
-    const trie = new CheckpointTrie({ db: new LevelDB() })
-    await trie.put(k1, v1)
-    t.deepEqual(await trie.get(k1), v1, 'before CP: v1')
-    trie.checkpoint()
-    await trie.put(k1, v12)
-    trie.checkpoint()
-    await trie.put(k1, v123)
-    await trie.commit()
-    t.deepEqual(await trie.get(k1), v123, 'after commit (second CP): v123')
-    await trie.revert()
-    t.deepEqual(await trie.get(k1), v1, 'after revert (first CP): v1')
-    t.end()
-  })
-})
+import { BatchDBOp, MemoryDB, LevelDB } from '../src/db'
+
+for (const DB of [MemoryDB, LevelDB]) {
+  tape('testing checkpoints', function (tester) {
+    const it = tester.test
+
+    let trie: CheckpointTrie
+    let trieCopy: CheckpointTrie
+    let preRoot: String
+    let postRoot: String
+
+    it('setup', async function (t) {
+      trie = new CheckpointTrie({ db: new DB() })
```

Comment:
Sorry, it will likely be in the code and easy to read out, but just (also) to have the explicit conversation: what happens when this generic `DB` class is used? Does this default to something in particular? The memory DB?

Then I wonder why we are not defaulting to this in the first place since this would be also some kind of implicit defaulting? 🤔

Or don't I get the whole picture?

Reply:
> DB in this case here is just an alias for the concrete implementation that is being tested

Ah, thanks didn't see the broader context.

> We could default to the MemoryDB if that's what you prefer.

I think I do have some preference on defaulting to MemoryDB after thinking about it twice, sorry for the extra work this undecisiveness might eventually cause. My feeling is that this implicit behavior is just "very innocent" - by not using any DB at all and just stay "in code". It is very unlikely that this behavior would every change (to - e.g. - always start with a full on-disk level DB or something) and for using an on-disk DB one can use the option explicitly. And at the same time things get a bit more innocent if people just want to play around with the Trie library without having the need to always look up this `db` constructor usage.

But let's wait, I might be unnecessarily strong opinionated, and eventually in a wrong direction. I will post this within the team chat and collect some feedback from the team. Will come back here if there is some strong tendency in the one or the other direction.

Reply:
I reached out to the team and have unanimous feedback that they want to keep a default value here to make things simpler for beginners / experimental usages.

Can you readopt? Sorry that this has caused some extra work.